### DISCUSSION TITLE
ValueError: [E177] Ill-formed IOB input detected

### DISCUSSION THREAD

**Discussion by sbushmanov** [explosion/spaCy#5753] [Help: Coding & Implementations] [answered]
When I try co convert simple conllu file ([link](https://github.com/natasha/nerus) to project page, [link](https://storage.yandexcloud.net/natasha-nerus/data/nerus_lenta.conllu.gz) to full corpus, 2Gb) with ner the following error appears:

```
spacy convert -c 'ner' try.conllu > try.json

Traceback (most recent call last):
  File "/home/sergey/anaconda3/envs/rasa/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/sergey/anaconda3/envs/rasa/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/sergey/anaconda3/envs/rasa/lib/python3.7/site-packages/spacy/__main__.py", line 33, in <module>
    plac.call(commands[command], sys.argv[1:])
  File "/home/sergey/anaconda3/envs/rasa/lib/python3.7/site-packages/plac_core.py", line 328, in call
    cmd, result = parser.consume(arglist)
  File "/home/sergey/anaconda3/envs/rasa/lib/python3.7/site-packages/plac_core.py", line 207, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "/home/sergey/anaconda3/envs/rasa/lib/python3.7/site-packages/spacy/cli/convert.py", line 106, in convert
    no_print=no_print,
  File "/home/sergey/anaconda3/envs/rasa/lib/python3.7/site-packages/spacy/cli/converters/conll_ner2json.py", line 111, in conll_ner2json
    biluo_ents = iob_to_biluo(iob_ents)
  File "gold.pyx", line 570, in spacy.gold.iob_to_biluo
  File "gold.pyx", line 592, in spacy.gold._consume_ent
ValueError: [E177] Ill-formed IOB input detected: =
```
The simplified source file is:
```
# newdoc id = 0
# sent_id = 0_0
# text = Вице-премьер по социальным вопросам Татьяна Голикова рассказала, в каких регионах России зафиксирована наиболее высокая смертность от рака, сообщает РИА Новости.
1	Вице-премьер	_	NOUN	_	Animacy=Anim|Case=Nom|Gender=Masc|Number=Sing	7	nsubj	_	Tag=O
2	по	_	ADP	_	_	4	case	_	Tag=O
3	социальным	_	ADJ	_	Case=Dat|Degree=Pos|Number=Plur	4	amod	_	Tag=O
4	вопросам	_	NOUN	_	Animacy=Inan|Case=Dat|Gender=Masc|Number=Plur	1	nmod	_	Tag=O
5	Татьяна	_	PROPN	_	Animacy=Anim|Case=Nom|Gender=Fem|Number=Sing	1	appos	_	Tag=B-PER
6	Голикова	_	PROPN	_	Animacy=Anim|Case=Nom|Gender=Fem|Number=Sing	5	flat:name	_	Tag=I-PER
```
Ner tags are given in the form `Tag=O` ,`Tag=B-PER`,  `Tag=I-PER` etc.

It would be really helpful if:
-  you provide some description as to what input file format is expected (perhaps, with examples)
-  what `ner` input format is expected.

if the first most probably is [CoNLL-U](https://universaldependencies.org/format.html), it would be helpful to explicitly state so, with a link to the format definition. The second needs an explicit specification.

Any help with this error is appreciated.

## Which page or section is this issue related to?

https://spacy.io/api/cli#convert

**Comment by adrianeboyd** [comment]
I think you're accidentally using the wrong converter here. `-c ner` doesn't support CoNLL-U format. The converter docs are here: https://spacy.io/api/cli#convert with links to the examples for the supported NER formats with `-c ner` here: https://github.com/explosion/spaCy/tree/27a1cd3c630055802513a20c1e75d0b37943cc39/examples/training/ner_example_data 

For CoNLL-U you can leave off the `-c` option (it checks by file ending) or use `-c conllu`. The problem with CoNLL-U is that there is no standard format for NER annotation. The converter is intended to convert most of the standard CoNLL-U columns (currently in spacy v2 this is columns FORM/UPOS/XPOS/HEAD/DEP). For treebanks that we use in some of the provided models, it also supports a few formats for including NER in the MISC column, but none of these are an official supported part of CoNLL-U or our converter.

Let's see, I think the simplest option with spacy v2 is to remove the `Tag=` to have plain NER tags in the MISC column. If your data had `SpaceAfter=No` or other features in the MISC column the spacy v2 converter isn't going to work correctly, but if MISC is just plain NER tags, I think the converter will detect them correctly. I usually recommend the following command to group sentences into small paragraphs for better training (this is mainly so the parser learns to split sentences and you may want to experiment with it for your data):

```
spacy convert -n 10 file.conllu .
```

If you need support for `SpaceAfter=No` or other features in MISC, let me know and I can explain how to use the converter from the current alpha version of spacy v3. (I wouldn't recommend using the alpha v3 version for training yet, but the converter itself is stable and has a lot of improvements over v2. I used it internally for all the updated v2.3 models. It supports more NER formats in MISC, all of the CoNLL-U columns, and `SpaceAfter=No` in the MISC column to create raw texts.)

**Comment by sbushmanov** [comment]
Thanks a lot for a quick response!

1. `spacy convert file.conllu .` works. `spacy convert -n 3 file.conllu .` works too. `spacy convert -n 10 file.conllu .` produces empty file.

2. In the json output I have unesaped unicode: `"\u0422\u0430\u0442\u044c\u044f\u043d\u0430"`. Is there a way to have readable unicode text (like in `json.dumps(...., ensure_ascii=False)`)?
 
Any chance to get help with a related issue while training?

`ValueError: [E069] Invalid gold-standard parse tree. Found cycle between word IDs: {59, 51, 54}`

(Strangely enough, I'm just training `ner`:

```
spacy train ru /home/sergey/Py_Spacy_RU/test_del /home/sergey/Py_Spacy_RU/data/nerus/nerus_json/try.json /home/sergey/Py_Spacy_RU/data/nerus/nerus_json/try.json --base-model /home/sergey/Py_Spacy_RU/ru2 --n-iter 20 --n-early-stopping 5 --pipeline 'ner'
```
)

Is there a way to have a list of misbehaving sentences (ids, or just starting text like in 2.3.1), so that sentences with cycles be removed at once? Or maybe you may advise a different way of treating cycles like ignoring them?

**Comment by adrianeboyd** [comment]
> Thanks a lot for a quick response!
> 
>     1. `spacy convert file.conllu .` works. `spacy convert -n 3 file.conllu .` works too. `spacy convert -n 10 file.conllu .` produces empty file.

That is pretty unexpected behavior for the `-n` option. I'd check that your output contains the same number of sentences and the same texts as the input in case it's due to a problem with the input and something more serious is going wrong.
 
>     2. In the json output I have unesaped unicode: `"\u0422\u0430\u0442\u044c\u044f\u043d\u0430"`. Is there a way to have readable unicode text (like in `json.dumps(...., ensure_ascii=False)`)?

No, not directly with `spacy convert`, we just don't have an option for this built in. You'd have to read in the file with `json` and export it again with the `ensure_ascii` option to reformat it.

> Any chance to get help with a related issue while training?
> 
> `ValueError: [E069] Invalid gold-standard parse tree. Found cycle between word IDs: {59, 51, 54}`
> 
> (Strangely enough, I'm just training `ner`:
> 
> ```
> spacy train ru /home/sergey/Py_Spacy_RU/test_del /home/sergey/Py_Spacy_RU/data/nerus/nerus_json/try.json /home/sergey/Py_Spacy_RU/data/nerus/nerus_json/try.json --base-model /home/sergey/Py_Spacy_RU/ru2 --n-iter 20 --n-early-stopping 5 --pipeline 'ner'
> ```
> 
> )
> 
> Is there a way to have a list of misbehaving sentences (ids, or just starting text like in 2.3.1), so that sentences with cycles be removed at once? Or maybe you may advise a different way of treating cycles like ignoring them?

Hmm, I think it tries to read in all the available annotation in case you might be training a parser later. I would just remove the "dep" and "head" values for all the tokens from the corpus in the JSON files to get around this if you're only training an NER model.

**Comment by sbushmanov** [comment]
> That is pretty unexpected behavior for the -n option. I'd check that your output contains the same number of sentences and the same texts as the input in case it's due to a problem with the input and something more serious is going wrong.

Source file:

```
(rasa) sergey@sergey-Bionic:~/Py_Spacy_RU/data/nerus$ cat try_.conllu
# newdoc id = 0
# sent_id = 0_0
# text = Вице-премьер по социальным вопросам Татьяна Голикова рассказала, в каких регионах России зафиксирована наиболее высокая смертность от рака, сообщает РИА Новости.
1	Вице-премьер	_	NOUN	_	Animacy=Anim|Case=Nom|Gender=Masc|Number=Sing	7	nsubj	_	O
2	по	_	ADP	_	_	4	case	_	O
3	социальным	_	ADJ	_	Case=Dat|Degree=Pos|Number=Plur	4	amod	_	O
4	вопросам	_	NOUN	_	Animacy=Inan|Case=Dat|Gender=Masc|Number=Plur	1	nmod	_	O
5	Татьяна	_	PROPN	_	Animacy=Anim|Case=Nom|Gender=Fem|Number=Sing	1	appos	_	B-PER
6	Голикова	_	PROPN	_	Animacy=Anim|Case=Nom|Gender=Fem|Number=Sing	5	flat:name	_	I-PER
7	рассказала	_	VERB	_	Aspect=Perf|Gender=Fem|Mood=Ind|Number=Sing|Tense=Past|VerbForm=Fin|Voice=Act	0	root	_	O
8	,	_	PUNCT	_	_	13	punct	_	O
9	в	_	ADP	_	_	11	case	_	O
10	каких	_	DET	_	Case=Loc|Number=Plur	11	det	_	O
11	регионах	_	NOUN	_	Animacy=Inan|Case=Loc|Gender=Masc|Number=Plur	13	obl	_	O
12	России	_	PROPN	_	Animacy=Inan|Case=Gen|Gender=Fem|Number=Sing	11	nmod	_	B-LOC
13	зафиксирована	_	VERB	_	Aspect=Perf|Gender=Fem|Number=Sing|Tense=Past|Variant=Short|VerbForm=Part|Voice=Pass	7ccomp	_	O
14	наиболее	_	ADV	_	Degree=Pos	15	advmod	_	O
15	высокая	_	ADJ	_	Case=Nom|Degree=Pos|Gender=Fem|Number=Sing	16	amod	_	O
16	смертность	_	NOUN	_	Animacy=Inan|Case=Nom|Gender=Fem|Number=Sing	13	nsubj:pass	_	O
17	от	_	ADP	_	_	18	case	_	O
18	рака	_	NOUN	_	Animacy=Inan|Case=Gen|Gender=Masc|Number=Sing	16	nmod	_	O
19	,	_	PUNCT	_	_	20	punct	_	O
20	сообщает	_	VERB	_	Aspect=Imp|Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin|Voice=Act	0	root	_	O
21	РИА	_	PROPN	_	Animacy=Inan|Case=Nom|Gender=Neut|Number=Sing	20	nsubj	_	B-ORG
22	Новости	_	PROPN	_	Animacy=Inan|Case=Nom|Gender=Fem|Number=Plur	21	appos	_	I-ORG
23	.	_	PUNCT	_	_	20	punct	_	O

# sent_id = 0_1
# text = По словам Голиковой, чаще всего онкологические заболевания становились причиной смерти в Псковской, Тверской, Тульской и Орловской областях, а также в Севастополе.
1	По	_	ADP	_	_	2	case	_	O
2	словам	_	NOUN	_	Animacy=Inan|Case=Dat|Gender=Neut|Number=Plur	9	parataxis	_	O
3	Голиковой	_	PROPN	_	Animacy=Anim|Case=Gen|Gender=Fem|Number=Sing	2	nmod	_	B-PER
4	,	_	PUNCT	_	_	2	punct	_	O
5	чаще	_	ADV	_	Degree=Cmp	9	advmod	_	O
6	всего	_	PRON	_	Animacy=Inan|Case=Gen|Gender=Neut|Number=Sing	5	obl	_	O
7	онкологические	_	ADJ	_	Case=Nom|Degree=Pos|Number=Plur	8	amod	_	O
8	заболевания	_	NOUN	_	Animacy=Inan|Case=Nom|Gender=Neut|Number=Plur	9	nsubj	_	O
9	становились	_	VERB	_	Aspect=Imp|Mood=Ind|Number=Plur|Tense=Past|VerbForm=Fin|Voice=Mid	0	root	_O
10	причиной	_	NOUN	_	Animacy=Inan|Case=Ins|Gender=Fem|Number=Sing	9	xcomp	_	O
11	смерти	_	NOUN	_	Animacy=Inan|Case=Gen|Gender=Fem|Number=Sing	10	nmod	_	O
12	в	_	ADP	_	_	20	case	_	O
13	Псковской	_	ADJ	_	Case=Loc|Degree=Pos|Gender=Fem|Number=Sing	20	amod	_	B-LOC
14	,	_	PUNCT	_	_	15	punct	_	O
15	Тверской	_	ADJ	_	Case=Loc|Degree=Pos|Gender=Fem|Number=Sing	13	conj	_	B-LOC
16	,	_	PUNCT	_	_	17	punct	_	O
17	Тульской	_	ADJ	_	Case=Loc|Degree=Pos|Gender=Fem|Number=Sing	13	conj	_	B-LOC
18	и	_	CCONJ	_	_	19	cc	_	O
19	Орловской	_	ADJ	_	Case=Loc|Degree=Pos|Gender=Fem|Number=Sing	13	conj	_	B-LOC
20	областях	_	NOUN	_	Animacy=Inan|Case=Loc|Gender=Fem|Number=Plur	9	obl	_	I-LOC
21	,	_	PUNCT	_	_	25	punct	_	O
22	а	_	CCONJ	_	_	25	cc	_	O
23	также	_	ADV	_	Degree=Pos	22	fixed	_	O
24	в	_	ADP	_	_	25	case	_	O
25	Севастополе	_	PROPN	_	Animacy=Inan|Case=Loc|Gender=Masc|Number=Sing	20	conj	_	B-LOC
26	.	_	PUNCT	_	_	9	punct	_	O

# sent_id = 0_2
# text = Вице-премьер напомнила, что главные факторы смертности в России — рак и болезни системы кровообращения.
1	Вице-премьер	_	NOUN	_	Animacy=Anim|Case=Nom|Gender=Masc|Number=Sing	2	nsubj	_	O
2	напомнила	_	VERB	_	Aspect=Perf|Gender=Fem|Mood=Ind|Number=Sing|Tense=Past|VerbForm=Fin|Voice=Act	0	root	_	O
3	,	_	PUNCT	_	_	11	punct	_	O
4	что	_	SCONJ	_	_	11	mark	_	O
5	главные	_	ADJ	_	Case=Nom|Degree=Pos|Number=Plur	6	amod	_	O
6	факторы	_	NOUN	_	Animacy=Inan|Case=Nom|Gender=Masc|Number=Plur	11	nsubj	_	O
7	смертности	_	NOUN	_	Animacy=Inan|Case=Gen|Gender=Fem|Number=Sing	6	nmod	_	O
8	в	_	ADP	_	_	9	case	_	O
9	России	_	PROPN	_	Animacy=Inan|Case=Loc|Gender=Fem|Number=Sing	7	nmod	_	B-LOC
10	—	_	PUNCT	_	_	11	punct	_	O
11	рак	_	NOUN	_	Animacy=Inan|Case=Nom|Gender=Masc|Number=Sing	6	nsubj	_	O
12	и	_	CCONJ	_	_	13	cc	_	O
13	болезни	_	NOUN	_	Animacy=Inan|Case=Nom|Gender=Fem|Number=Plur	11	conj	_	O
14	системы	_	NOUN	_	Animacy=Inan|Case=Gen|Gender=Fem|Number=Sing	13	nmod	_	O
15	кровообращения	_	NOUN	_	Animacy=Inan|Case=Gen|Gender=Neut|Number=Sing	14	nmod	_	O
16	.	_	PUNCT	_	_	2	punct	_	O

# sent_id = 0_3
# text = В начале года стало известно, что смертность от онкологических заболеваний среди россиян снизилась впервые за три года.
1	В	_	ADP	_	_	2	case	_	O
2	начале	_	NOUN	_	Animacy=Inan|Case=Loc|Gender=Neut|Number=Sing	4	obl	_	O
3	года	_	NOUN	_	Animacy=Inan|Case=Gen|Gender=Masc|Number=Sing	2	nmod	_	O
4	стало	_	VERB	_	Aspect=Perf|Gender=Neut|Mood=Ind|Number=Sing|Tense=Past|VerbForm=Fin|Voice=Act	0	root	_O
5	известно	_	ADJ	_	Degree=Pos|Gender=Neut|Number=Sing|Variant=Short	4	xcomp	_	O
6	,	_	PUNCT	_	_	14	punct	_	O
7	что	_	SCONJ	_	_	14	mark	_	O
8	смертность	_	NOUN	_	Animacy=Inan|Case=Nom|Gender=Fem|Number=Sing	14	nsubj	_	O
9	от	_	ADP	_	_	11	case	_	O
10	онкологических	_	ADJ	_	Case=Gen|Degree=Pos|Number=Plur	11	amod	_	O
11	заболеваний	_	NOUN	_	Animacy=Inan|Case=Gen|Gender=Neut|Number=Plur	8	nmod	_	O
12	среди	_	ADP	_	_	13	case	_	O
13	россиян	_	NOUN	_	Animacy=Anim|Case=Gen|Gender=Masc|Number=Plur	8	nmod	_	O
14	снизилась	_	VERB	_	Aspect=Perf|Gender=Fem|Mood=Ind|Number=Sing|Tense=Past|VerbForm=Fin|Voice=Mid	4	ccomp	_	O
15	впервые	_	ADV	_	Degree=Pos	14	advmod	_	O
16	за	_	ADP	_	_	18	case	_	O
17	три	_	NUM	_	Animacy=Inan|Case=Acc	18	nummod:gov	_	O
18	года	_	NOUN	_	Animacy=Inan|Case=Gen|Gender=Masc|Number=Sing	15	obl	_	O
19	.	_	PUNCT	_	_	4	punct	_	O

```
Convert command:

```
(rasa) sergey@sergey-Bionic:~/Py_Spacy_RU/data/nerus$ spacy convert -n 10 try_.conllu .
✔ Generated output file (0 documents)
try_.json

```
Output:
```
(rasa) sergey@sergey-Bionic:~/Py_Spacy_RU/data/nerus$ cat try_.json 
[

]
```
Thanks in advance!

**Comment by adrianeboyd** [comment, marked as answer]
I'm not sure what version of spacy you're using? You don't have to use it for training, but try the latest v2 (currently v2.3.1) for converting. I get the error:

```
File "/path/spacy/cli/converters/conllu2json.py", line 69, in read_conllx
    id_, word, lemma, pos, tag, morph, head, dep, _1, iob = parts
ValueError: not enough values to unpack (expected 10, got 9)
```

This line is malformed:

```none
13	зафиксирована	_	VERB	_	Aspect=Perf|Gender=Fem|Number=Sing|Tense=Past|Variant=Short|VerbForm=Part|Voice=Pass	7ccomp	_	O
```

I suspect your best option is to use tools from universal dependencies to validate the CoNLL-U format.

**Comment by sbushmanov** [comment]
`spacy convert -n 10 file.conllu .` works indeed in the latest spacy (2.3.1).

The empty file I mentioned was due to spacy 2.1.8.
